pkg/reconciler/revision/revision.go

```diff
@@ -97,53 +98,31 @@ func (c *Reconciler) reconcileDigest(ctx context.Context, rev *v1.Revision) erro
 	}
 
 	var digestGrp errgroup.Group
-	type digestData struct {
-		digestValue        string
-		containerName      string
-		isServingContainer bool
-		image              string
-		digestError        error
-	}
-
-	digests := make(chan digestData, len(rev.Spec.Containers))
-	for _, container := range rev.Spec.Containers {
+	rev.Status.ContainerStatuses = make([]v1.ContainerStatuses, len(rev.Spec.Containers))
+	for i, container := range rev.Spec.Containers {
 		container := container // Standard Go concurrency pattern.
+		i := i
 		digestGrp.Go(func() error {
 			digest, err := c.resolver.Resolve(container.Image,
 				opt, cfgs.Deployment.RegistriesSkippingTagResolving)
 			if err != nil {
-				err = fmt.Errorf("failed to resolve image to digest: %w", err)
-				digests <- digestData{
-					image:       container.Image,
-					digestError: err,
-				}
+				return errors.New(v1.RevisionContainerMissingMessage(container.Image, fmt.Sprintf("failed to resolve image to digest: %v", err)))
 			} else {
-				isServingContainer := len(rev.Spec.Containers) == 1 || len(container.Ports) != 0
-				digests <- digestData{
-					digestValue:        digest,
-					containerName:      container.Name,
-					isServingContainer: isServingContainer,
+				if len(rev.Spec.Containers) == 1 || len(container.Ports) != 0 {
+					rev.Status.DeprecatedImageDigest = digest
+				}
+				rev.Status.ContainerStatuses[i] = v1.ContainerStatuses{
+					Name:        container.Name,
+					ImageDigest: digest,
 				}
 			}
 			return nil
 		})
 	}
-	digestGrp.Wait()
-	close(digests)
-	for v := range digests {
-		if v.digestError != nil {
-			rev.Status.MarkContainerHealthyFalse(v1.ReasonContainerMissing,
-				v1.RevisionContainerMissingMessage(
-					v.image, v.digestError.Error()))
-			return v.digestError
-		}
-		if v.isServingContainer {
-			rev.Status.DeprecatedImageDigest = v.digestValue
-		}
-		rev.Status.ContainerStatuses = append(rev.Status.ContainerStatuses, v1.ContainerStatuses{
-			Name:        v.containerName,
-			ImageDigest: v.digestValue,
-		})
+	if err := digestGrp.Wait(); err != nil {
+		rev.Status.MarkContainerHealthyFalse(v1.ReasonContainerMissing, err.Error())
+		rev.Status.ContainerStatuses = make([]v1.ContainerStatuses, 0)
```

Comment:
why do we need to do this?

Reply:
The `TestContainerErrorMsg` e2e test was failing because if we don't clear this out, we will try to create image cache so the config will fail with a different message https://github.com/knative/serving/blob/f2d5f674c695c3df91bdc76f82d74c5c2e8f5024/pkg/reconciler/revision/reconcile_resources.go#L119

Reply:
1. why not nil?
2. you can do this differently by using local variable for the list and only assigning it if `err != nil`.